packages/api/src/beacon/client/events.ts

```diff
@@ -1,56 +1,52 @@
-import {stringifyQuery, urlJoin} from "../../utils/client/format.js";
+import {ChainForkConfig} from "@lodestar/config";
 import {getEventSource} from "../../utils/client/eventSource.js";
-import {HttpStatusCode} from "../../utils/client/httpStatusCode.js";
+import {stringifyQuery, urlJoin} from "../../utils/client/format.js";
+import {ApiClientMethods} from "../../utils/client/method.js";
+import {RouteDefinitionExtra} from "../../utils/client/request.js";
+import {ApiResponse} from "../../utils/client/response.js";
+import {BeaconEvent, Endpoints, definitions, getEventSerdes} from "../routes/events.js";
 
 /**
  * REST HTTP client for events routes
  */
-export function getClient(baseUrl: string): Api {
+export function getClient(_config: ChainForkConfig, baseUrl: string): ApiClientMethods<Endpoints> {
   const eventSerdes = getEventSerdes();
 
   return {
-    eventstream: async (topics, signal, onEvent) => {
+    eventstream: async ({topics, signal, onEvent, onError, onClose}) => {
       const query = stringifyQuery({topics});
-      const url = `${urlJoin(baseUrl, routesData.eventstream.url)}?${query}`;
+      const url = `${urlJoin(baseUrl, definitions.eventstream.url)}?${query}`;
       // eslint-disable-next-line @typescript-eslint/naming-convention
       const EventSource = await getEventSource();
       const eventSource = new EventSource(url);
 
-      try {
-        await new Promise<void>((resolve, reject) => {
-          for (const topic of topics) {
-            eventSource.addEventListener(topic, ((event: MessageEvent) => {
-              const message = eventSerdes.fromJson(topic, JSON.parse(event.data));
-              onEvent({type: topic, message} as BeaconEvent);
-            }) as EventListener);
-          }
-
-          // EventSource will try to reconnect always on all errors
-          // `eventSource.onerror` events are informative but don't indicate the EventSource closed
-          // The only way to abort the connection from the client is via eventSource.close()
-          eventSource.onerror = function onerror(err) {
-            const errEs = err as unknown as EventSourceError;
-            // Consider 400 and 500 status errors unrecoverable, close the eventsource
-            if (errEs.status === 400) {
-              reject(Error(`400 Invalid topics: ${errEs.message}`));
-            }
-            if (errEs.status === 500) {
-              reject(Error(`500 Internal Server Error: ${errEs.message}`));
-            }
-
-            // TODO: else log the error somewhere
-            // console.log("eventstream client error", errEs);
-          };
-
-          // And abort resolve the promise so finally {} eventSource.close()
-          signal.addEventListener("abort", () => resolve(), {once: true});
-        });
-      } finally {
+      const close = (): void => {
         eventSource.close();
+        onClose?.();
+        signal.removeEventListener("abort", close);
```

Comment:
As it's one time listener, so need to remove. This way you can use this listener as inline similars to all others used in this file.

Reply:
Are you suggestion to just remove the line 
```ts
signal.removeEventListener("abort", close);
```
or anything else in addition to that?

Reply:
I noticed that `close` is used on multiple places, so that's fine.